helpers/embeds.js

```diff
@@ -0,0 +1,82 @@
+import { html, render } from 'lit';
+import { findComposedAncestor } from '../helpers/dom.js';
+import { tryGetPluginByKey } from '@brightspace-ui/core/helpers/plugins.js';
+
+const embedTypeAttributeName = 'data-d2l-embed-type';
+
+class RenderNode {
+	#childNodes;
+	#node;
+
+	constructor(node) {
+		this.#node = node;
+		this.#childNodes = [];
+	}
+
+	addChildNode(childNode) {
+		this.#childNodes.push(childNode);
+	}
+
+	async renderNode(options) {
+		await Promise.all(this.#childNodes.map(childNode => childNode.renderNode(options)));
+
+		const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', this.#node.dataset.d2lEmbedType);
+		if (!embedRendererPlugin) return;
+
+		const fragment = new DocumentFragment();
+		this.#node.childNodes.forEach(node => {
+			const temp = node.cloneNode(true);
+			fragment.appendChild(temp);
+		});
+
+		let props = this.#node.dataset.d2lEmbedData;
+		try {
+			props = JSON.parse(this.#node.dataset.d2lEmbedData);
+		} catch (e) { /* empty */ }
+
+		const embed = await embedRendererPlugin.renderView(fragment, props || {}, options || {});
+		if (embed) {
+			render(embed, this.#node.parentNode);
+			this.#node.remove();
+		}
+	}
+
+}
+
+export async function createEmbedPlaceholder(embedType, options) {
+	const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', embedType);
+	if (!embedRendererPlugin) return;
+
+	const placeholderData = await embedRendererPlugin.getPlaceholder(options);
+	/* NOT IMPLEMENTED */
+}
+
+export async function renderEmbeds(node, options) {
+	const elems = node.querySelectorAll(`div[${embedTypeAttributeName}], span[${embedTypeAttributeName}]`);
+	if (elems.length === 0) return;
+
+	const renderNodes = new Map();
+	const rootNodes = [];
+	const constructRenderNode = elem => {
+		if (renderNodes.has(elem)) return;
+
+		const renderNode = new RenderNode(elem);
+		renderNodes.set(elem, renderNode);
+
+		const isEmbedAncestor = ancestor => {
+			if (ancestor === elem || ancestor.nodeType !== Node.ELEMENT_NODE) return;
+			return ancestor.hasAttribute(embedTypeAttributeName);
+		};
+
+		const nearestEmbedAncestor = findComposedAncestor(elem, isEmbedAncestor);
+		if (!nearestEmbedAncestor) {
+			rootNodes.push(renderNode);
+			return;
+		}
+
+		renderNodes.get(nearestEmbedAncestor).addChildNode(renderNode);
+	};
+
+	elems.forEach(constructRenderNode);
```

Comment:
Perhaps more of a stylistic thing, but it may be possible to write this in such a way that the helper method is more stateless, such that it doesn't mutate things outside itself:
```javascript
const getRenderNodeContext = elem => {
    ...
    const renderNode = new RenderNode(elem);
    ...
    return {
        ancestorElem: nearestEmbedAncestor,
        renderNode: renderNode
    };
};

elems.forEach(elem => {
    const context = getRenderNodeContext(elem);
    renderNodes.set(elem, context.renderNode);
    context.ancestorElem ? 
        renderNodes.get(context.ancestorElem).addChildNode(context.renderNode) : 
        rootNodes.push(context.renderNode);
});
```

In your current implementation, `renderNode` is always added to `renderNodes` (via `set()`), and then it would potentially also be added to a `renderNode`'s child array (via `addChildNode`). As such, it would potentially be referenced twice in the `renderNodes` "tree". Is that intentional?

I think the naming of things is kinda confusing right now. Whenever I read "node"/"Node", I think of an actual DOM node. Like `renderNode` sounds like a method that renders an actual DOM node. I think I would consider renaming to avoid that confusion. It would also be good to make it clearer that the elems are the embed placeholders - future readers trying to understand this code might need to read a bunch of the code to discern that they aren't the replacement elements. I'd probably also be intentional about "node" vs "elem", but that's less confusing. I don't want to bike shed on naming, but straightening that out might help.

Reply:
Yeah, so as written right now, this duplication exists because we're using the `renderNodes` map solely to build up the array in `rootNodes`. Basically, `rootNodes` ends up being an array of placeholders we need to process, each with children (with their own children, etc.). i.e. Those are the trees we care about. `renderNodes` is just a temporary map to make sure we don't try to process a placeholder multiple times, so if an element has, say, 10 children, we'll add it to the tree when we find the first child and then know to skip it for the next nine. It's possible there are cleaner ways to do this, but a map stuck out to me at first because of it's quick insertion/lookup time.

RE: the nomenclature and general structure, this is just something quick and dirty I threw together to prove out that this is possible, so I haven't put much thought into structure or where state is being tracked. I fully expect that most of this will be heavily rewritten to be cleaner when we have a firm idea that this (or something else) is the direction we want to take. Unfortunately that does mean a lot of repeated `elem` and `elems` names in various scopes. If it makes it easier to parse for now, I can change up the names to be more obvious, but I don't want to spend a ton of time thinking about it until we know for sure we're not going to swing in some wildly different direction, if that makes sense.